### DISCUSSION TITLE
New Signal Proposal : Sawtooth Wave

### DISCUSSION THREAD

**Discussion by OwenDavid** [Hopsan/hopsan#2043] [Show and tell]
The sawtooth wave is a typical signal source and can be found as standard in other modelling software, such as Amesim and Simster from BoschRexroth.

The attached code is a proposal for a new signal component which outputs a sawtooth wave train. The component is based on the pulse wave from the default library.

I trust this will be a useful addition and worthy of consideration

[SignalSawtoothWave.zip](https://github.com/Hopsan/hopsan/files/5790736/SignalSawtoothWave.zip)

**Comment by robbr48** [comment]
Good idea, this has been missing a long time in Hopsan. The implementation seems correct, but I am concerned about the inexact maximum values. The signal sometimes reaches the maximum value a little too early, and then remains constant a short while before falling. I am not sure if it is due to the use of floating-point numbers for counter variables, comparing floating-point numbers or just an integration error. A more accurate solution would be to use a modulo operator with some additional computations. Here is my attempt:
```
double y0 = *mpBaseValue;
double T = *mpPeriodT;
double D = *mpDutyCycle;
double yA = *mpAmplitude;
double T0 = *mpStartTime;
if(mTime < T0) {
    (*mpOut) = y0;
}
else {
    (*mpOut) = y0 + fmax(0.0,(fmod(mTime+T0+T*(1-D),T)/T-(1.0-D))/D)*yA;
}
```
What do you think?
